File: milestone 2/python_modules/module02/ex4/ft_raise_errors.cpp
#include <iostream>
#include <string>
#include <cstring>
using namespace std;

class ValueError : public std::exception
{
    public:
    string name, desc;
    
    ValueError(const string &name, const string &desc)
    : name(name), desc(desc) {}

    const char* what() const noexcept override
    {
        return desc.c_str();
    }
};

int check_plant_health(string plant_name, int water_level, int sunlight_hours)
{
    try
    {
        if (plant_name.empty())
            throw ValueError("NameError", "Plant name cannot be empty!");
    }
    catch(const std::exception& e)
    {
        std::cerr << "Error: " << e.what() << '\n';
        return 1;
    }

    try
    {
        if (water_level <= 1)
            throw ValueError("ValueError", "low");
        else if (water_level >= 10)
            throw ValueError("ValueError", "high");
    }
    catch(const std::exception& e)
    {
        if (strcmp(e.what(), "low") == 0)
            std::cerr << "Error: Water level " << water_level << " is too low (min 1)" << '\n';
        else if (strcmp(e.what(), "high") == 0)
            std::cerr << "Error: Water level " << water_level << " is too high (max 10)" << '\n';
    }

    try
    {
        if (sunlight_hours <= 2)
            throw ValueError("ValueError", "cold");
        else if (sunlight_hours >= 12)
            throw ValueError("ValueError", "hot");
    }
    catch(const std::exception& e)
    {
        if (strcmp(e.what(), "cold") == 0)
            std::cerr << "Error: Sunlight hours " << sunlight_hours << " is too low (min 2)" << '\n';
        else if (strcmp(e.what(), "hot") == 0)
            std::cerr << "Error: Sunlight hours " << sunlight_hours << " is too high (max 12)" << '\n';
    }
    
    if ((!plant_name.empty()) && (water_level >= 1 && water_level <= 10) && (sunlight_hours >= 2 && sunlight_hours <= 12))
    {
        cout << "Plant '" << plant_name << "' is healthy!" << endl; 
    }
    return 0;
}

void test_plant_checks(void)
{
    cout << "\nTesting good values...\n";
    check_plant_health("tomato", 3, 8);
    cout << "\nTesting empty plant name...\n";
    check_plant_health("", 2, 1);
    cout << "\nTesting bad water level...\n";
    check_plant_health("tomato", 15, 4);
    cout << "\nTesting bad sunlight hours...\n";
    check_plant_health("tomato", 4, 0);
}

int main(void)
{
    cout << "=== Garden Plant Health Checker ===\n";
    test_plant_checks();
    cout << "\nAll error raising tests completed!" << endl;
}

File: milestone 2/python_modules/module02/ex4/ft_raise_errors.py
def check_plant_health(plant_name: str, water_level: int, sunlight_hours: int):
    try:
        if not plant_name:
            raise ValueError("Plant name cannot be empty!")
    except ValueError as e:
        print(f"Error: {e}")
        return

    try:
        if water_level > 10:
            raise ValueError(
            f"Water level {water_level} is too high (max 10)")
        elif water_level < 1:
            raise ValueError(
            f"Water level {water_level} too low (min 1)")
    except ValueError as e:
        print(f"Error: {e}")

    try:
        if sunlight_hours > 12:
            raise ValueError(
                f"Sunlight hours {sunlight_hours} is too high (max 12)")
        elif sunlight_hours < 2:
            raise ValueError(
                f"Sunlight hours {sunlight_hours} is too low (min 2)")
    except ValueError as e:
        print(f"Error: {e}")

    if ((water_level <= 10 and water_level >= 1)
    and (sunlight_hours <= 12 and sunlight_hours >= 2)
    and (plant_name is not None)):
        print(f"Plant '{plant_name}' is healthy!")


def test_plant_checks():
    print("\nTesting good values...")
    check_plant_health("tomato", 4, 8)
    print("\nTesting empty plant name...")
    check_plant_health("", 8, 4)
    print("\nTesting bad water level...")
    check_plant_health("apple", 15, 6)
    print("\nTesting bad sunlight hours...")
    check_plant_health("rosemary", 7, 0)


if __name__ == "__main__":
    print("=== Garden Plant Health Checker ===")
    test_plant_checks()
    print("\nAll error raising tests completed!")